### DISCUSSION TITLE
Automating HAZUS loss and damage calculations using R2D

### DISCUSSION THREAD

**Discussion by digant98** [NHERI-SimCenter/SimCenterCommon#289] [R2D]
I'm working on a project to calculate flood damage and losses for buildings in my study area. I currently have flood depth grids and want to streamline the damage calculation process using HAZUS. A peer recommended looking into the R2D application as part of the workflow. Is there a script-based version of R2D for HAZUS that can be accessed via Python or R scripts? Or, anyone has experience automating HAZUS calculations in a similar setup?

**Comment by zsarnoczay** [comment]
Thank you for reaching out! 

There are several ways you can run your assessment with SimCenter tools:

- R2D provides a user interface and a built-in QGIS to help you review inputs and outputs and manage the simulation.

- You can run the exact same simulation if you prepare your inputs in a folder and directly feed them to the SimCenter's simulation engine. R2D does the same, it takes the inputs you specify in the GUI, copies them to a folder and then prepares a configuration file that specifies which file is used for which application in the simulation. If you are interested in this route, I recommend running R2D at least once, then using the files it prepared as a template, modifying them and running the backend directly from then on.

- Since your simulation is rather simple, a third option is to run only the damage and loss calculation using our Pelicun tool. This tool is part of our larger simulation engine, but it is also available as a standalone Python package. It allows file-based operation where you prepare input files and run a simulation (this is used in our large workflows) and also allows interactive operation where you import the package in a python script and use its modules to perform the calculations as part of a bigger workflow or a Jupyter notebook. You might be interested in this if you already have a python-based workflow and only need the damage and loss analysis step.

We can help you with all three options above, just let us know which one is most appealing to you.

Let me also ask you a few questions on your plans:

- You mentioned that you are interested in damage calculation. The flood model in Hazus uses loss functions to estimate the loss ratio (i.e., the proportion of building value lost) as a function of inundation height relative to the first floor of the building. We have these functions available in our [Damage and Loss Model Library](https://github.com/NHERI-SimCenter/DamageAndLossModelLibrary/tree/main/flood/building/portfolio/Hazus%20v6.1). Would such loss functions work for your project?

- There are hundreds of archetypes available in Hazus (see the number of rows in the `loss_repair.csv` file in the above link) and it is not always trivial to map the right archetype to buildings in your inventory. Do you have a building inventory available or were you planning to use Hazus' building inventory? What type of buildings are in your study area? Don't worry if you don't have a building inventory yet - we can help.

- How are your flood depth grids stored? Are they in raster files? What file format do they use?

- What kind of outputs are you looking for? For example: aggregate expected losses for the study area or the entire probabilistic distribution of losses for each building?

**Reply by jinyan1214** [reply]
Thanks @digant98 for your interest, and @zsarnoczay thanks a lot for the detailed answer!

To add more resources, the R2D documentation also has[ some instructions](https://nheri-simcenter.github.io/R2D-Documentation/common/developer_manual/architecture/desktop/run-manually/runLocal.html) on running R2D through Python.

**Reply by digant98** [reply]
@zsarnoczay thanks so much for such detailed answer and willingness to help further! 

I think the third method - run only the damage and loss calculation using our Pelicun tool could be helpful. As I mentioned earlier, we already have python workflow to create flood depth grid. 

Here are the answers to your specific questions: 

1) _You mentioned that you are interested in damage calculation. The flood model in Hazus uses loss functions to estimate the loss ratio (i.e., the proportion of building value lost) as a function of inundation height relative to the first floor of the building. We have these functions available in our [Damage and Loss Model Library](https://github.com/NHERI-SimCenter/DamageAndLossModelLibrary/tree/main/flood/building/portfolio/Hazus%20v6.1). Would such loss functions work for your project?_
**Answer**: Yes, the functions you have available in Damage and Loss Model Library works for our project.

2) _There are hundreds of archetypes available in Hazus (see the number of rows in the loss_repair.csv file in the above link) and it is not always trivial to map the right archetype to buildings in your inventory. Do you have a building inventory available or were you planning to use Hazus' building inventory? What type of buildings are in your study area? Don't worry if you don't have a building inventory yet - we can help._
**Answer**: We are planning to use HAZUS level 1 inventory data available through HAZUS. Our study area is pretty large to collect detailed information about each individual buildings. 

3) _How are your flood depth grids stored? Are they in raster files? What file format do they use?_
**Answer**: Our flood depth grids are stored as a raster files.

4) _What kind of outputs are you looking for? For example: aggregate expected losses for the study area or the entire probabilistic distribution of losses for each building?_
**Answer**: As I mentioned our study area is large, there are hundreds of buildings in the area. Therefore, aggregate expected losses for the study area is the best way forward.

I appreciate your willingness to help us! @zsarnoczay  @jinyan1214 
I look forward to your response.

**Reply by zsarnoczay** [reply]
Thanks for the responses, we can proceed with a Pelicun-based approach. 

Let me remind you about a few caveats with this approach:
- You will need to bring your own building inventory and prepare the Pelicun input files in the standard format we use.
- You will need to use your flood depth grids and map it to your building locations to figure out the inundation height for each building.

I assume you were already planning to do the above steps yourself, but you might be interested to know that R2D and the underlying simulation engine include functionality that can automate these tasks. So, if you are not particularly excited about them, you might want to take a look at the other options.

Let me ask a few further questions to clarify some additional details:

- Are you planning to extract the inventory data from Hazus and prepare your building inventory yourself? The Hazus building inventory is provided as an aggregate number of buildings per occupancy type and year of construction in each census block. You will need to infer additional building characteristics (such as first floor elevation, foundation type) to be able to identify appropriate flood vulnerability archetypes for the buildings. Some of our tools can help you with these steps if you need such help.

- Do you plan to develop a building inventory that identifies individual structures? The alternative is to keep buildings geographically linked to census blocks and not to worry about where exactly they are located within the block. Given your interest in flooding, I assume you'd be inclined to have a higher geographical resolution that census block because the flood inundation heights tend to change rapidly in space.

- Are you familiar with Jupyter notebooks? The functionality you are looking for is relatively new in Pelicun and we do not have a Pelicun-specific example online. I plan to create an example Jupyter notebook that shows you how to run such an analysis through Pelicun. Would that be a convenient way for you to learn more about these features?

**Reply by digant98** [reply]
@zsarnoczay Thanks for your response! 

Regarding the caveats with the Pelicun-based approach: I am already planning to use flood depth grids and map it to your building locations to figure out the inundation height for each building. I'll appreciate more information about preparing the Pelicun input files in the standard format you use for the building inventory. It's good to know R2D and the underlying simulation engine include functionality that can automate these tasks!

Answers to your specific questions:

- _Are you planning to extract the inventory data from Hazus and prepare your building inventory yourself? The Hazus building inventory is provided as an aggregate number of buildings per occupancy type and year of construction in each census block. You will need to infer additional building characteristics (such as first floor elevation, foundation type) to be able to identify appropriate flood vulnerability archetypes for the buildings. Some of our tools can help you with these steps if you need such help._
**Answer**: We have a shapefile with polygons indicating actual building location, occupancy type, and corresponding total building height for ~12000 structures in our study area (this data comes from FEMA's USA Structures State GDB Download Site - https://disasters.geoplatform.gov/USA_Structures/). Problem with this data is that it does not contain information about foundation type, floor specific height information, etc. Do you have tool that can help with these? 

- _Do you plan to develop a building inventory that identifies individual structures? The alternative is to keep buildings geographically linked to census blocks and not to worry about where exactly they are located within the block. Given your interest in flooding, I assume you'd be inclined to have a higher geographical resolution that census block because the flood inundation heights tend to change rapidly in space._
**Answer**: Yes, I would prefer to develop a building inventory that identifies individual structures using above mentioned building inventory shapefile data (FEMA's USA Structures State GDB).  

- _Are you familiar with Jupyter notebooks? The functionality you are looking for is relatively new in Pelicun and we do not have a Pelicun-specific example online. I plan to create an example Jupyter notebook that shows you how to run such an analysis through Pelicun. Would that be a convenient way for you to learn more about these features?_
**Answer**: Yes, I'm familiar with Jupyter notebook. An example Jupyter notebook that shows how to run such an analysis through Pelicun would be very convenient. Thank you so much!
